database.py

```diff
@@ -561,6 +561,18 @@ def mark_submission_invalid(self, submission_id: SubmissionId, /) -> None:
 
             self.refresh_user_best_runs(year, day, part, user)
 
+    def pick_new_container_versions(self, versions: list[str]) -> set[str]:
+        """
+        Given a list of container versions, return those not in the DB already.
+        """
+        query = "SELECT container_version FROM container_versions"
+        given_versions = set(versions)
+        res = self._cursor.execute(query).fetchall()
+        if res is None or len(res) == 0:
+            return set(versions)
+        present_versions = set(map(lambda x: x[0], res))
+        return given_versions.difference(present_versions)
```

Comment:
I was going to suggest using a set-expr here for increased perf, but its too much of a nitpick to matter, but then i saw the next line, we are storing two sets in memory unnecessarily since we immediately discard them after one computation
this can just become:
```suggestion
        for x in res:
            try:
                given_versions.remove(x[0])
            except KeyError:
                pass
        return given_versions
```

it can also replace the `if res is None or len(res) == 0` check, since if the length is empty we will remove nothing from the set, and just return it directly

also, 
`Return all (remaining) rows of a query result as a [list](https://docs.python.org/3/library/stdtypes.html#list). Return an empty list if no rows are available. Note that the [arraysize](https://docs.python.org/3/library/sqlite3.html#sqlite3.Cursor.arraysize) attribute can affect the performance of this operation.`

fetchall wont return None, and computes to a list unnecessarily, we really dont need to call it and then we will only have the memory use of the initial set passed in, that gets subtracted from as we iterate the cursor directly, minimizing computation and memory use

Reply:
Honestly, this query is stupid. We've got a paged HTTP API used to specify a list, and we need to get the DB to remove items it's already seen in a paged manner. The actual correct way would probably be inserting pages of tags into a temp table, executing a query to delete rows from the temp table, and then reading out the temp table a batch at a time. But that's a lot of extra code for an already huge PR that's blocking development, and we don't need it right now, afaik.